modules/core/shared/src/main/scala-3/io/circe/derivation/ConfiguredEncoder.scala

```diff
@@ -0,0 +1,44 @@
+package io.circe.derivation
+
+import scala.deriving.Mirror
+import scala.compiletime.constValue
+import io.circe.{Encoder, Json, JsonObject}
+
+trait ConfiguredEncoder[A](using conf: Configuration) extends Encoder.AsObject[A], DerivedInstance[A]:
```

Comment:
Why does the trait have public methods `elemEncoders`, `encodeProduct` and `encodeSum`?

Reply:
I saw no reason to hide these methods.

Reply:
Public interface is something you cannot change after release as that would break binary compatibility. Did you foresee any usage of those methods outside of this file?

Reply:
And having both `final def encodeProduct(a: A): JsonObject` and `final def encodeSum(index: Int, a: A): JsonObject` does not make sense.

Reply:
There are more alternatives to making the methods either public and final or private. We could add an annotation like mayChange, internal, experimental, etc. These annotations would imply these methods can be used, but the library is still free to break the binary compatibility, aka, use them at your own risk.

I see no problem in having both `encodeProduct` and `encodeSum` be final. Can you give more details?

I foresee a slight possibility of someone wanting to customize the derivation in a way that is not supported by the current configuration, and not wanting to re-implement most methods from `ConfiguredEncoder`/`ConfiguredDecoder`, just because they are private.

Reply:
Also the `derived` methods on `Configured`* classes return the ConfiguredEncoder/Decoder/Codec explicitly to more easily enable the use of the methods from those classes (like `encodeProduct` or `decodeSum`)

Reply:
> I see no problem in having both encodeProduct and encodeSum be final. Can you give more details?

Sorry for not being clear, my note was not about them being final but rather about their presence in general. Suppose I have a type `A` for which I get a derived instance of `ConfiguredEncoder`. `A` is either sum or product, not both. However it seems I am allowed to call both `encodeProduct` and `encodeSum`. I would rather keep the `ConfiguredEncoder` as minimal as possible. It could serve one distinct purpose: ability to express that one wants a configured codec e.i codec that requires `Configuration` in scope. Methods that you think could be reused may be put in another trait (which you can mix-in in the implementation).

Reply:
I see what you are getting at, but the trait approach would not work. How can you have the methods public in the trait, but make them private when you mix them in?

Lack of experience/knowledge from the programmer invoking the methods is a weak reason not to expose them. We could make the derived method transparent and return a more specialized type (think `ConfiguredSumDecoder` and `ConfiguredProductDecoder`). Those types could then have the appropriate method. Personally I don't think is worth the extra complexity (extra classes, the more confusing derived method, etc), however the implementation is easy.

Reply:
> How can you have the methods public in the trait, but make them private when you mix them in?
You can mix them in `derived` method where you actually provide implementation.

> Lack of experience/knowledge from the programmer invoking the methods is a weak reason not to expose them. 
I believe we should aim for typesafety here. Invoking these methods will throw exceptions at runtime in certain cases which is not ideal for an interface designed for everyday use.